File: java/leetcode_grind/Day580.java
package leetcode_grind;

public class Day580 {
    // https://leetcode.com/problems/sum-of-square-numbers/description/
    static class Solution1 {
        public boolean judgeSquareSum(int c) {
            for (int i = 2; i * i <= c; i++) {
                int count = 0;
                if (c % i == 0) {
                    while (c % i == 0) {
                        count++;
                        c /= i;
                    }
                    if (i % 4 == 3 && count % 2 != 0) {
                        return false;
                    }
                }
            }
            return c % 4 != 3;
        }
    }
}

File: src/leetcode_grind/day_580.rs
// https://leetcode.com/problems/sum-of-square-numbers/description
pub fn judge_square_sum(c: i32) -> bool {
    let mut a = 0;
    while a * a <= c {
        let mut b = c as i64 - (a as i64 * a as i64);

        let mut lo = 0i64;
        let mut hi = b;

        while lo <= hi {
            let mid = lo + (hi - lo) / 2;
            if mid * mid == b {
                return true;
            }
            if mid * mid < b {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        a += 1;
    }
    false
}
